app/src/main/java/com/vci/vectorcamapp/main/presentation/MainViewModel.kt

```diff
@@ -64,18 +67,21 @@ class MainViewModel @Inject constructor(
         }
     }
 
-    private fun determineStartDestination() {
+    private fun observeDevice() {
         viewModelScope.launch {
-            try {
-                val device = deviceCache.getDevice()
-                _state.update {
-                    it.copy(
-                        startDestination = if (device == null) Destination.Registration else Destination.Landing
-                    )
+            deviceCache.getDeviceFlow()
+                .catch {
+                    emit(null)
+                    emitError(MainError.DEVICE_FETCH_FAILED)
                 }
-            } catch (e: Exception) {
-                emitError(MainError.DEVICE_FETCH_FAILED)
-            }
+                .onEach { device ->
+                    if (device != null && _state.value.startDestination != Destination.Landing) {
+                        _state.update { it.copy(startDestination = Destination.Landing) }
+                    } else if (device == null && _state.value.startDestination != Destination.Registration) {
+                        _state.update { it.copy(startDestination = Destination.Registration) }
+                    }
+                }
+                .collect { }
```

Comment:
I wonder if you can do something along the lines of how we load the incomplete and complete sessions, and then combine it with the _state using the combine function. See if something like that is possible. That might help get rid of this observeDevice function altogether.